### ISSUE TITLE
Flash Display

### ISSUE THREAD

**Issue by hydroper** [ROBERT-MCDOWELL/Flash#2]
The rs ecosystem is just growing. We can search [crates.io](https://crates.io/categories/rendering) for finding what we need from what we had in Flash, except the display tree (`flash.display`). With [gfx-rs](https://github.com/gfx-rs/gfx), winit and swf-rs we can implement a good one.

Porting old projects to Rust might be hard since it's gradually different from ActionScript, though. But that's for good since AS isn't a scalable language.

**Comment by ROBERT-MCDOWELL** [comment]
I changed the git repo name to be better found on git search, but didn't know it just deleted all discussion issues
crates.io looks good, as long as it will be better than flash.display....

**Comment by ROBERT-MCDOWELL** [comment]
I also recalled FutureFlashFramework by Foo Flash Framework.
I want to keep the work Flash to link Flash developers and user interest.
btw Flash is not copyrighted anymore since 2015

**Comment by ROBERT-MCDOWELL** [comment]
> Porting old projects to Rust might be hard since it's gradually different from ActionScript, though. 
but feasible?

**Comment by ROBERT-MCDOWELL** [comment]
first issue is always good to keep, yes I just saw it, it's ok to leave it as this.
sorry, I wanted to say gfx-rs, not crates.io
> Yeah. Not much hard, but the developer is encouraged to change the code design.
ok, but for the SWF without sources it should work right?

**Comment by hydroper** [comment]
Yes. We don't have much to do, just need to learn math and get used to low-level graphics, swapbuffers... Or use high-level graphics from Amethyst, SDL2, Idk.

There's no multi-purpose display tree implemented by anyone in Rust, still. A UDP server has been implemented by the looks: https://github.com/christian-smith/rust-udp-server.

Well, if we want port SWFs easily then it's a means of having a `flash` crate. It won't cost in anything in Rust.

**Comment by hydroper** [comment]
Amethyst is for ECD (Entity Component System) patterns, so it's mostly for games. Flash is for ECS, Functional Reactive and all other graphical programming patterns.

**Comment by hydroper** [comment]
> Amtthyst looks promising, but why a pitty? :\

Since it's ECS-based, that means it's for game-purposes only. But not much to worry, I'm writting the lib. It'll allow you to cast between the sealed classes (`DisplayObject`, `Sprite`, `Shape`, `MovieClip`, `TextField` etc.) just like you could in AS, but you won't be able to extend them (since that mostly leads to bad code), at least not properly.

The cool thing is that will be usable for ECS too. It'll be usable for everything, UI, these.

`objects/mod.rs`

```
mod unit;
mod inherit;
mod disp_obj;
mod sprite;

pub use unit::*;
pub use inherit::InstanceOf;
pub use disp_obj::DisplayObject;
pub use sprite::*;
```

`objects/disp.rs`

```
use fladisp::*;
use std::convert::*; // AsRef, From, Into

pub struct DisplayObject {
    kind: DispObjKind,
    x: Unit,
    y: Unit,
}

// Impl InstanceOf's
```

**Comment by hydroper** [comment]
> udp server looks good, so an RTMFP protocol can for sure be created modifying this project code.

Hmm, we could implement a Flash equivalent into Rust, then. E.g., a `flash::net` module that supports it.

**Comment by hydroper** [comment]
To get used to Rust you can check _Rust by Example_.
https://doc.rust-lang.org/stable/rust-by-example/ 

Well, as ActionScript developers we've used Flash Professional/Animate or MXML directly to build projects. Rust has a package manager in command-line and it's not that hard to use it.

After installing Rust you can check:

https://doc.rust-lang.org/book/ch01-03-hello-cargo.html

It'll introduce you, even though it's, yeah, different from ActionScript. It's really light.

**Comment by hydroper** [comment]
Ah, you _do_ install [rustup](https://rustup.rs) with the `night` channel. For now it's the one that stdweb supports.

**Comment by ROBERT-MCDOWELL** [comment]
ok will do. We need also to think about a flexible and smart architecture.
Let's say if Rust becomes a dead project, what are the chance to switch to another environment to avoid to not loose all the work? From the past lessons we always need to adapt the future to avoid to repeat the same mistakes

**Comment by hydroper** [comment]
@ROBERT-MCDOWELL The framework doesn't depend in many particularities of Rust. But if it's not done in Rust, then it means we'll re-duplicate a low-level system to target WebAssembly, nope?

As to the display library (which I'm now having to work in phone...), as planned, it could be adapted to ActionScript/C#/Swift sealed class inheritance. Inheritance is important since it allows cheap field access through display elements. With that you can perform collision checks with various objects more efficiently (by first capturing rectangular collisions and then exact collisions).

Rust becoming a dead project is unlikely. Very different technologies may come (quantum computing), but binary ones will continue to still be exponentially faster for gaming, animating and such. SIMD operations, for instance, are already in std. They aim to be growing.

Rust is super unlikely to die due to its transparency. When something is transparent, you've full control and confidency of what you're doing.

Swift is also kinda there, but it looks like it had reflection by what's said out. I've never rant it since I've no opportunity to that so fine.

Rust is a snake, orochi.

**Comment by ROBERT-MCDOWELL** [comment]
- The framework doesn't depend in many particularities of Rust. But if it's not done in Rust, then it means we'll re-duplicate a low-level system to target WebAssembly, nope?
I don't know, my question was from my ignorance. I even don't know if it's possible. What's the most long term language existing if it's not C and C++? so if Rust discontinues, does any conversion tools allow use to convert Rust language to C/C++? it's questions we need to seriously ask.
- Rust is super unlikely to die due to its transparency
mhhm, well, I won't say the story is repeating every decade, but it's unfortunately what I see through the internet development. Everything is nice and promising when a new stuff is created and many developers use it, then boom, one day the core staff decides to stop it for various reasons. I saw that for many languages/projects/softwares, Flash included. What you said about Rust above was exactly the same words 10 years ago with Flash. So it's always better to prevent rather than loose years of work.

**Comment by ROBERT-MCDOWELL** [comment]
- The framework doesn't depend in many particularities of Rust. But if it's not done in Rust, then it means we'll re-duplicate a low-level system to target WebAssembly, nope?
 I don't know, my question was from my ignorance. I even don't know if it's possible. What's the most long term language existing if it's not C and C++? so if Rust discontinues, does any conversion tools allow us to convert Rust language to C/C++? it's questions we need to seriously ask.
- Rust is super unlikely to die due to its transparency
mhhm, well, I won't say the story is repeating every decade, but it's unfortunately what I see through the internet development. Everything is nice and promising when a new stuff is created and many developers use it, then boom, one day the core staff decides to stop it for various reasons. I saw that for many languages/projects/softwares, Flash included. What you said about Rust above was exactly the same words 10 years ago with Flash. So it's always better to prevent rather than loose years of work.

**Comment by hydroper** [comment]
@ROBERT-MCDOWELL But Flash isn't transparent, that's when we include its runtime. It requires many assumptions (e.g., store all API strings in ABC constant pool and bring objects with lots of properties representing the API classes at run-time, which can then also be seen by external SWFs and thus not optimised ahead-of-time).

C, C++ and Rust are the kind of transparent languages. Ah, and Rust wasn't implemented in C++, nor there's a Rust-2-C\[++] transpiler. Rust is a product of lessions taught earlier from C++ by Mozilla.

What you say about projects stopping didn't exactly occur with, say, C++.

**Comment by ROBERT-MCDOWELL** [comment]
- What you said about projects stopping didn't exactly occur with, say, C++.
Ok so we need to know if, for the worst case scenario there is a bridge between Rust-2-C++ and C++ right?

**Comment by ROBERT-MCDOWELL** [comment]
To start such kind of big project we always need to think of all the worst scenario and negative aspects of it.

**Comment by ROBERT-MCDOWELL** [comment]
what do you think of this article?
https://hackernoon.com/why-im-dropping-rust-fd1c32986c88

**Comment by ROBERT-MCDOWELL** [comment]
and this one?
https://www.viva64.com/en/b/0324/

**Comment by hydroper** [comment]
@ROBERT-MCDOWELL They're outdated. I've read the Hackernoon one and its main issue is inheritance. Inheritance _is_ possible (it was my first issue, too, since I were to implement display/DOM-like objects). After having used Rust more, all needed is implementing conversion traits (`From`, `TryFrom` and `AsRef`), fallback traits (`Deref`), a custom `InstanceOf` trait and unsafe casts internally (`Rc::from_raw(Rc::into_raw(r.clone()) as *const() as *const Out)`). 

I'm writting a macro `class_decl!` to facilitate that. Inheritance is efficient, but still the Rust folks don't get it right. They think I want delegation with traits or composition; err, I can't really communicate with them well.

Ah, they do many RFCs, but the language got better in 2018. It's really different from C++, e.g., there's no formal spec. exactly, but in the other hand the progress is fast in it.

Now, Viva64's article looks to also criticise the package manager. What? It's saying packages that depend in other packages hosted in GitHub will cause crates.io to be crazyland. In first place, a package _cannot_ be published to crates.io when it has _external_ dependencies; it can depend in crates.io itself, but cannot in GitHub or a local file-system. And to add, the Rust ecosystem is well aware of semvers (**s**emantic **v**ersions), so it's quite the opposite of what this article thinks.

Overall they leak experience with Node.JS, Cargo and inheritance.

**Comment by hydroper** [comment]
It didn't go well. Well, still inheritance is hard in need of zero-cost `Deref`. Hmm, I'm starting to think Rust isn't a good idea, then.

**Comment by hydroper** [comment]
Entity-Component-System isn't something I want directly; I'd rather use it in top of `flash.display`, thus escaping from inheritance to composition or trait delegation isn't appropriate.

**Comment by hydroper** [comment]
N't sure, but I'm thinking a ActionScript alternative has to come later. E.g. a namespaced-language inspired by AS3. I was working in such, but since I gave up misc. times I won't be working in it any time soon. I guess we'll have to use AS3 as is, but breaking compatibility with the spec (and thus indexing and run-time reflection would be broken).

**Comment by ROBERT-MCDOWELL** [comment]
I found this 
https://internals.rust-lang.org/t/why-not-inheritance/5738/21
maybe it will help, if not, le'ts continue to dig in the rabbit hole

**Comment by hydroper** [comment]
@ROBERT-MCDOWELL They've still not added a feature like _fields in traits_. I've done a class_decl, but I couldn't implement `Deref` efficiently and am thus not satisfied w/ Rust. What a pitty...

**Comment by hydroper** [comment]
Everything in Rust is good, except `Rc`. Lalaolohuahu

**Comment by hydroper** [comment]
I could do something different for the standard's `Rc`, but why, really? Hm...

**Comment by ROBERT-MCDOWELL** [comment]
@hydroper it it helps to get inheritance.... What's the motivation of the authors of Rust to lack of it?

**Comment by hydroper** [comment]
They look to say it's not a priority, but is really affecting the flexibility of future libs, I think so.

**Comment by ROBERT-MCDOWELL** [comment]
weird. anyhow, it stays C++ right?

**Comment by hydroper** [comment]
Yes. C++ isn't very unsafe (compared to Rust it's unsafer), but I don't like the usual build systems (Make, CMake, Bazel and others). Wauh. No module system... :(

**Comment by hydroper** [comment]
Okay, I better don't spam about inheritance here. I'll still give a try to replacing `Rc` somehow.

**Comment by ROBERT-MCDOWELL** [comment]
There is always a way to do it, it only needs the righ idea ;)

**Comment by ROBERT-MCDOWELL** [comment]
https://github.com/WebAssembly/design/issues/1255#issuecomment-455003301

**Comment by hydroper** [comment]
Crystal has inheritance, but its host/target systems are still very undone. Well, meanwhile only C++ has inheritance and targets systems well (like Rust).

Inheritance is something efficient/flexible for display objects, but what I need isn't inheritance at all; I just need efficient field traits within reference-counted structures. The unique way will be to diverge from Rust standard library's `Rc` pointer, but I think I'll just use Adobe AIR for now. I'm writting a package manager for ActionScript and maybe something good might happen.

**Comment by hydroper** [comment]
What I see is that Adobe AIR has few-few showcases. I know apps/games done w/ Adobe AIR (Atelier 801 games (principally the newest, Dead Maze, released in 2017), Hero Fighter X, Brawhalhia, Gasha Universe and I'm not reminding right now other samples). 3D games there're too, but all of which I know aren't MMO (I even remind a nearly 3D RPG).

**Comment by ROBERT-MCDOWELL** [comment]
In digital world the beauty is everything is possible, from bad to good. It just needs imagination, out of all squared guides, manual and else...

**Comment by hydroper** [comment]
I solved to continue building a refinement language of ActionScript. (Now that I've deleted my unicode-gc repo and lost its source, I'm so sick about it...)

  * [Vega in \~7 mins.](https://github.com/vega-lang/vega-show)
  * [W.I.P. semantics](https://github.com/vega-lang/vegac/blob/master/semantics.md)

**Comment by hydroper** [comment]
Vega isn't exactly a subset of AS. It's not AS, but by the syntax and `namespace`s it looks like.

**Comment by ROBERT-MCDOWELL** [comment]
ok good idea, but is the purpose of this new language to be used with Rust or?

**Comment by hydroper** [comment]
@ROBERT-MCDOWELL Mostly with C++. If it were Rust, then I'd have to waste neurons with the borrow checker and general crates obligging me to borrow several times.

**Comment by ROBERT-MCDOWELL** [comment]
ok gotcha. So what the advantages to create a new language on top of C++ knowing that for ex
Haxe is an Actionscript like language and can be compiled in native?

**Comment by hydroper** [comment]
@ROBERT-MCDOWELL Haxe has RTT-info (reflection and property names in run-time), what means it already has certain overhead for large projects. But whether that RTT-info would cause slowness, well, it depends in AOT optimizations. If dead code isn't eliminated, much may happen.

Haxe is at least better than ActionScript, but it's not entirely near a systems programming language. Its implementation uses a mark-sweep GC. If you take into account mark-sweep GC and RTT-info, it has a quite considerable overhead for the Web.

RTT-info? Well:

```js
class A {
    constructor() {
        this.f = 'Hi'
    }
}

const a = new A
a[String.fromCodePoint('b'.charCodeAt(0) + 4)]
```

**Comment by hydroper** [comment]
@ROBERT-MCDOWELL Pros:

  * Flexibler names (with `namespace`s) and aliasing versus Haxe/AS
  * Really better `import`s (and the compiler aims to allow better directory-structure)
  * Data-classes (`Vector3F` as a copy-value everywhere)
  * Many syntax goodies (`function` keyword only used in non-method definitions; lambdas use arrow-likes)
  * Implementation will use A.R.C. (auto ref-count) rather than tracing GC since the language provides `WeakRef.<>`
  * Faster-to-write loops (numeric ones may use (inclusive) ranges, e.g., `for (var i in 1..=j) task;`, though there's no step like in Lua)
  * Safer `switch`, safer in general.

Cons:

  * No reflection (definitions can't be hooked at run-time; properties aren't exposed t oo).

**Comment by hydroper** [comment]
Error-handling is also different. It's Rust-based, but discrete.

**Comment by ROBERT-MCDOWELL** [comment]
ok, Do you think the Cons are the only major one? Could it be fixed? because properties are really important for more flexibility

**Comment by ROBERT-MCDOWELL** [comment]
What the alternative?

**Comment by hydroper** [comment]
@Robert-McDoWell It's the equivalent of writting maps/dictionaries with string-keys. In languages like Rust and C++ there's usually no reflection too.

**Comment by hydroper** [comment]
Sometimes specific error handling is neccessary for calling functions, so `Function` is of no value, but if it's there it'll do nothing since dead code elimination will be successful.

**Comment by hydroper** [comment]
Unlike other languages, Vega'd not need a passive JIT optimizator like V8.

**Comment by ROBERT-MCDOWELL** [comment]
how long it will need to finalize it?
do you know kha? 
https://tracker.adobe.com/#/view/FP-4198746

**Comment by ROBERT-MCDOWELL** [comment]
how long it will take to finalize this language?
do you know kha?
https://github.com/Kode/Kha

**Comment by hydroper** [comment]
@Robert-McDoWell Kha is a SDK, but looks promising for me. Lots of targets and supports Haxe directly.

Hmm, really interesting.

**Comment by hydroper** [comment]
As far as I know, Haxe doesn't target wasm anyway...

**Comment by ROBERT-MCDOWELL** [comment]
https://groups.google.com/forum/#!msg/haxelang/Pcm38LPFjW0/Q3-WXk2oBQAJ

**Comment by hydroper** [comment]
Haxe compiler also performs [DCE](https://haxe.org/manual/cr-dce.html). Personally I don't like to have Emscripten (implies JDK) installed in my environment; 'd be nicer if it were just LLVM (LLVM does support wasm today).

**Comment by hydroper** [comment]
But I know things can improve. Ehhr, I think I need to use Haxe.

**Comment by ROBERT-MCDOWELL** [comment]
Do you think that implement AVM1 is feasible with kha? as it's much smaller than AVM2 so it would be a good test.

**Comment by hydroper** [comment]
There're many built-ins in ActionScript, so there'll be a regular cost for any SWF running in Kha. Also doing it in Haxe won't be very efficient at all (but still better than in AS).

Conversion is the best way to fit existing projects.

**Comment by ROBERT-MCDOWELL** [comment]
There are many very good games and app not maintained anymore, or just with lost sources, I was wondering if there is a way to interpret teh SWF as it is through LLVM or such

**Comment by hydroper** [comment]
Yes, but you can already use such SWFs with standalone projectors... They won't be efficient for the Web, you could be sure, or unless they don't (accidentaly) use the so common AS reflection and external bytecode loading.

**Comment by hydroper** [comment]
Well, are you still referring to play SWFs in the Web?

**Comment by ROBERT-MCDOWELL** [comment]
yes, usually they download various things, not possible anymore with the legacy projector tech for security reasons

**Comment by hydroper** [comment]
Yes, I see. But it's not neccessary to use the Web for this implementation. Also, you can consider using old Firefox versions which allow plugins like Flash Player.

**Comment by hydroper** [comment]
(I heard about people using it.)

**Comment by hydroper** [comment]
E.g., download Firefox Portable.

**Comment by hydroper** [comment]
IMO what's missing is a SWF player for Android.

**Comment by ROBERT-MCDOWELL** [comment]
Users of my platform use https://www.puffin.com on Android and IOS, works very well

**Comment by ROBERT-MCDOWELL** [comment]
We constantly need to go back to the essential, find a way to run legacy and new standards on the same browser, that's why I created this repo...

**Comment by ROBERT-MCDOWELL** [comment]
https://thinkmobiles.com/blog/convert-flash-unity/

**Comment by hydroper** [comment]
Unity is mostly getting featured for ECS (entity component system). [Exploding Rabbit's Jay](https://explodingrabbit.com) is also re-writting his SMBC game in Unity/C# (with ECS) for use in a new upcoming game.

I think not everything in Flash has to be used in modern ways (like running any old SWF in your normal Chrome). As I said, everything will have to be emulated with pre-installed apps (Firefox Portable for MMO or SWF-external games in general, standalone Flash Player Debugger/Projector and standalone [PegaJogo](http://pegajogo.com)).

Only ellaborated SWFs which still sh

**Comment by hydroper** [comment]
I said miscellanous times about transparency here, too. Display objects, networking etc. are all fine, but due to AS (JS-subset) it's a bit inconvenient to fit SWFs in the web. Well, ... It has to also have more relationship to the Browsers API.

**Comment by hydroper** [comment]
Principally, one issue I had said earlier was SWFs that load external SWFs. Implement this feature for the Web is heavy as it also means embedding a WASM optimizer into the self Web page.

**Comment by hydroper** [comment]
Vega is a language for us used to AS, with static-everything (packages, namespaces, better code dependency through `[Require(src)]`). Being abstract and reference-counted-only it'll target all needs. It's like C++, but safer and a bit higher-level.

A scripting language like ActionScript that requires verification before all evaluations. In some senses it's like Swift (capture lists in lambdas for capturing locals as weak or unowned during reference-counting).

**Comment by hydroper** [comment]
(I preferred not to say about languages very much.)

**Comment by ROBERT-MCDOWELL** [comment]
-  It has to also have more relationship to the Browsers API to be a successful and costless project for bringing hand-heavy games there
that's all the challenge
I think about a game dev team I know running an online multiplayer Flash game since 2001 with 80 millions users. If we find a good solution I'm pretty sure there will be already a massive client to embrace the project as they didn't planned to reprogram a very well running game.

**Comment by ROBERT-MCDOWELL** [comment]
btw red tamarin, crossbridge essentially have been discontinued because of strategy choice, aka focus all their effort (Adobe included) to webassembly

**Comment by hydroper** [comment]
@ROBERT-MCDOWELL Hmmm...

> game dev team

Do you know which? Well, of the alive games I only know HF-X (inactive now) and A801 games in general...

**Comment by ROBERT-MCDOWELL** [comment]
I will tell you right on time when needed ;)

**Comment by hydroper** [comment]
There's much to do to get my language working. Hmm, I think I'm getting of my idea. Once I was able to get DollScript almost symbol-solving, but I did also waste time parsing it.

Well, I think Swift would be great really, but the it doesn't still target Windows very-well and WebAssembly. For me it'll be Haxe, err...

**Comment by ROBERT-MCDOWELL** [comment]
thing I know is today everything is done to make things very complex, but as it's done on purpose it means that there is always a narrow door opened we need to find. If it was easy so it would be already ready to use! :D

**Comment by ROBERT-MCDOWELL** [comment]
Ok  we are one day before to decide what to do to start.
My analyze and vision of things:
- all corpos controlling the web tech decided to adopt webassembly as the replacement of Flash and btw opening doors to multiple languages, as perspective get rid of JS which is obsolete, cpu and memory hunger, also hard to maintain. Of course webassembly will be the way for the major social corpomedias to improve all their JS spaghetti code. So the first point we can't avoid is webassembly, which on long term will even replace desktop softwares.
- Create a language is always possible, but on long term requires thousands of developers adopting it or for sure it will be dead before birth. To attract them, it needs a language that offers what other languages don't to make development to be easy, fast and rock solid. I don't have any doubt you are skillful to create this language, but something no one control is Time. Haxe, build on top of other languages is for me the most advanced Framework, in term of quality and popularity. I'm following this project since the start (12 years) and his Author, I can confirm that the quailty and results done on this project after 12 years is just brilliant, and a very important point, it's not controlled (for now) by any smoking corporations. it's genuine developers community from all around the world.
- So why not to not reinvent the wheel again and use what's already existing, and specialize/optimize it for what it's needed? explore deeper some part of Haxe to fill some gap/missing/features is not negligeable.

In conclusion, optimize Haxe, runtime machine for webassembly is the right solution for me, knowing that Nicolas Cannasse just started it, I guess that guys like you who join to the project will be nice. And at least you will be directly in the game, with no waste of time and doubts.

**Comment by hydroper** [comment]
> Of course webassembly will be the way for the major social corpomedias to improve all their JS spaghetti code. So the first point we can't avoid is webassembly, which on long term will even replace desktop softwares.

Haxe targets wasm by [transpilling to C++](https://github.com/HaxeFoundation/hxcpp), though. Maybe it were from this idea that they [started to think in reflection](https://haxe.org/manual/std-reflection.html). ;)

(A pitty the C++ runtime may include a mark-sweep GC. It could just use auto-ref.-counting if there wasn't the any-type and interface-casting limitations.)

> Haxe, build on top of other languages is for me the most advanced Framework, in term of quality and popularity.

Some perf. proofs are like the [Armory engine](https://armory3d.org), but frameworks may be missing networking. So, yay, we need contributing to existing projects or build something above them.

**Comment by ROBERT-MCDOWELL** [comment]
Isn't Kha be added as a module in Haxe?

**Comment by hydroper** [comment]
It might be in design, but they should adopt to the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API). For targetting HTML5 they'll want to reuse browser's `XMLHttpRequest` probably, but `fetch` is going everywhere.

**Comment by ROBERT-MCDOWELL** [comment]
or simply webrtc

**Comment by hydroper** [comment]
Yerr. Hm, I'm updating the source of ASC 2 (the ActionScript compiler by Adobe/Macromedia) to make a different edition of ActionScript. I won't have to parse the language, will just do some changes in its Java code and thus be able to bring AS in the web.

I'll have to plan a cooler symbol solver. I was thinking of transpilling Java to C# and use C# libraries for also doing different tasks such as doc generation (with Markdown), color-tracing to the console...

**Comment by hydroper** [comment]
I'm implementing a package manager for ActionScript with Node.js (I was about to use Rust, but I got some borrow limitations and got bored to do so; I was also about to use Red Tamarin or AIR, but none of them were appropriate enough for CLI apps and the former crashes arbitrarily).

**Comment by hydroper** [comment]
Anyway, right, I won't be able to work if I continue talking about the thrown-away situation.

> do you know asnodec?

Yes, but unstable: I've used AS2JS (from the same Apache Royale SDK), but I'm quite unable to install it here. It's looping infinitely with `npm install`.

I also like doing things in Node.js, except by the fact the module system support is crazyland.

**Comment by hydroper** [comment]
I'm already working in the package manager anyway. It'll kind of need `git` to make sure authors initialize new packages quickly.

**Comment by ROBERT-MCDOWELL** [comment]
ok right, it's always good to develop a better alternative. the golden rules being fast, easy, rock solid and clean for coming developers.... 
Yes please create you gir repo now.

**Comment by ROBERT-MCDOWELL** [comment]
What about SVG DOM v2? it can apparently embed code... don't know (yet) if javascript only

**Comment by hydroper** [comment]
@ROBERT-MCDOWELL I've never used the SVG API. As usual it's a [`HTMLElement`](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model#SVG_interfaces), but you use [`Document#createElementNS`](https://stackoverflow.com/questions/20539196/creating-svg-elements-dynamically-with-javascript-inside-html) to create it.

I've never trusted using SVG because, according to my talks with a experienced guy, it's rendered diagnostically to the device by browsers, but I'm not sure things have changed for SVG. I do rarely see SVG being used.

**Comment by ROBERT-MCDOWELL** [comment]
amazing how people like to reinvent the wheel over and over ;)
https://css-tricks.com/pong-svg-js/

**Comment by hydroper** [comment]
It looks like I'll get back to using Android in a portrait screen. In my house there's only 1 laptop now and I live in Brazil. I believe I'll be working mostly in Vega btw.

alien.as was totally erased in my SD card. It got corrupted because the wrapper flash drive keeps connecting and disconnecting in my laptop. :/

**Comment by hydroper** [comment]
I didn't commit things in GitHub nor did I sync my files in cloud.

**Comment by hydroper** [comment]
It wasn't hard to get writting Astro, but I got disappointed because after Node.js I wanted to do it in Red Tamarin, but the Red Tamarin SDK build for Linux is a corrupted ZIP!!!

**Comment by hydroper** [comment]
I said Linux because I'm in Android and there's a app _UserLAnd_ that runs Ubuntu. Meanwhile I can only use Dory.js. :/

**Comment by hydroper** [comment]
I'm using QuickEdit to program, but it's not a complete editor. It misses:
* extra key buttons,
* completion (buttons) and
* good highlighting.

But I'm able to write things in my phone anyway, luckily...

**Comment by hydroper** [comment]
So, it might take some time to bring Vega, but AS3 could be easily transitioned to Vega (once done), so no worries.

**Comment by hydroper** [comment]
To convert AS3 to Vega I think it's easier to inspect the code's bytecodes and generate approximate Vega code, requiring changes like `weak.<>` types etc. and improvements, such as avoiding nullable types.

**Comment by ROBERT-MCDOWELL** [comment]
- So, it might take some time to bring Vega, but AS3 could be easily transitioned to Vega (once done), so no worries.
the most important being the strongest and most efficient solution, so the time will follow

**Comment by ROBERT-MCDOWELL** [comment]
about Tamarin, is i this one?
https://sourceforge.net/projects/tamarin/files/tamarin/Tamarin%200.1/tamarin-0.1.zip/download?use_mirror=master&download=

**Comment by ROBERT-MCDOWELL** [comment]
Or this one
http://download.redtamarin.com/

**Comment by hydroper** [comment]
> Or this one
> http://download.redtamarin.com/

Yeah, sorry for missing links. Red Tamarin is actively maintained, but it summoned from the defunct Tamarin project. ⛪ I also know why I was experiencing crash issues; it was due to the AVM2 it embeds, which doesn't interpret ABC definition-order cleverly.

**Comment by ROBERT-MCDOWELL** [comment]
maybe crossbridge will be better so
https://github.com/crossbridge-community/crossbridge

**Comment by ROBERT-MCDOWELL** [comment]
http://old.haxe.org/com/libs/format/abc?version=19968

**Comment by hydroper** [comment]
I could hust use C++ for writting CLI, but I don't find C++ so rich. That `make` too...

In fact using `Makefile`s with `make` is at least better than using `mxmlc` in the command-line.

**Comment by hydroper** [comment]
Funny is, in Rust you just do `cargo new --bin hello-world` and in C++ you do `echo ......................... > Makefile`, but in Rust you don't program at all; instead you get compile-time errors forever.

In Swift things are different, but something like `unique_ptr` is missing too... It's also verbose and isn't distributed for every platform/machine (like Windows). Deceived for Apple...

**Comment by hydroper** [comment]
There're lots-lots of issues with Swift implementation, but so far C++ is more flexible. Didn't even think in that fact...

**Comment by hydroper** [comment]
There are problems with C++, still ― forgiviness of `null_ptr` checking, `shared_ptr` and `unique_ptr` and general-case built-ins (say, `Math`).

**Comment by hydroper** [comment]
Buuuuuuut... [C++ targets wasm32 (LLVM)](https://webassembly.org/docs/c-and-c++/), it targets anything. Just using `make` one is capable of much.

Vega will be an alternative language to C++. I'm not quite sure if using C++ would be an issue at all. I'd need to start using it in Termux... or UserLAnd. ;l

**Comment by hydroper** [comment]
https://www.reddit.com/r/cpp/comments/611811/have_you_used_rust_do_you_prefer_it_over_modern_c/ This somehow motivates me to continue on Rust, but I compiling it in Android is quite dangerous due to space, but I'm guessng it'd be great.

**Comment by ROBERT-MCDOWELL** [comment]
maybe rust is still immature for what you are trying to achieve?

**Comment by hydroper** [comment]
It's verbose, even for inheritance; out this it's powerful. Both stdweb and sdl2 fit binds inheritance in Rust, bringing `is`/`instanceof` operator as a `IsA`/`InstanceOf` trait, providing `From`` and `To` conversions.

**Comment by hydroper** [comment]
I don't think it's horrible, but compiling Rust code is hard because you'll always get errors, but when you fix them certainly (maybe) your code will be alright in run-time.

**Comment by hydroper** [comment]
For anyone that has logic in mind, I tend to think  `mut`and immitability restrictions like moving a borrow is very abusing. ---__---

**Comment by hydroper** [comment]
Somehow these borrow limitations might change program logic. Maybe lesss error-prone?

Well, even [Cone](https://github.com/jondgoodwin/cone) works like this.

**Comment by ROBERT-MCDOWELL** [comment]
if Cone works like this so maybe there is no other better solution then?

**Comment by hydroper** [comment]
A different solution is to not support _borrows_. Borrow types allow you to move _local_ variables to further scopes etc.. As you know, locals are variables created from program/function execution.

There's something similiar to borrows, like pass by-reference in PHP, C#, Swift and VB.

```php
function rotate(&$p) { $p = new Unimpl }

$p = new P; $a = $p;
rotate($p);
Assert::different($p, $a);
```

That doesn't run, but illustrates a bit the thing in PHP. It's in my opinion stupid to pollute a language with borrows.

Generic lifetime parameters in Rust/Cone allows `'a`-scope structures and flexible borrowing, but it causes confusion very generally and thus turns out Rust/Cone are time-wasting solutions... if you know you're gonna program right.

**Comment by hydroper** [comment]
I've probably overestimated. If you get used to programming with lifetime arguments, then it's truly fine...

I guess me myself I should pratice Rust more.

**Comment by hydroper** [comment]
> if Cone works like this so maybe there is no other better solution then?
Okay, I've overestimated Rust and Cone power. For me they're the best out there.

P.S.: Cone isn't completely implemented.

**Comment by ROBERT-MCDOWELL** [comment]
you mean there is better out there?

**Comment by hydroper** [comment]
Nope, edited.

**Comment by hydroper** [comment]
https://forge.rust-lang.org/platform-support.html

**Comment by hydroper** [comment]
The first thing I'll try creating in Rust... is an editor for my Android. ;)

**Comment by hydroper** [comment]
Well, btw, there's no toolchain for Android in `rustup`, there are only Android targets. Uhh...

I can't also install Android NDK due to space unless I put it in my corrupt SDCard. So I'll have to do so!

**Comment by ROBERT-MCDOWELL** [comment]
Do you need a new sdcard?

**Comment by hydroper** [comment]
Nah... It's working, but the laptop corrupts it.  `android_glue` API is incomplete... I'm not so used to using Ubuntu shell and it'd be hard to fork @tomaka repo for temporary usage.

In this case I'll have to also write a Android lib or I'll ignore the API and just use OpenGL ES from my Android for testing my Rust.

Hmm, we need to build something in Rust.

**Comment by ROBERT-MCDOWELL** [comment]
gosh...:\

**Comment by hydroper** [comment]
It's better to diverge from inspirations... this way. Aw. For me it ends here.

**Comment by ROBERT-MCDOWELL** [comment]
FYI do you know haxepunk?
https://github.com/HaxePunk/HaxePunk

**Comment by hydroper** [comment]
I could try Haxe on Ubuntu, but can't use OpenFL as well as .NET Core SDK in anyway. I could try the C++ backend.

I'm not sure if I'm inspired to program anything, and _yet-yet_ in _this_ Android environment.

**Comment by hydroper** [comment]
Unfortunately I can't program anymore. >.>
